github/resource_github_repository_environment.go

```diff
@@ -0,0 +1,257 @@
+	return &schema.Resource{
+		Create: resourceGithubRepositoryEnvironmentCreate,
+		Read:   resourceGithubRepositoryEnvironmentRead,
+		Update: resourceGithubRepositoryEnvironmentUpdate,
+		Delete: resourceGithubRepositoryEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"environment": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"wait_timer": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(0, 43200),
+			},
+			"reviewers": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 6,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"teams": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeInt},
+						},
+						"users": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeInt},
+						},
+					},
+				},
+			},
+			"deployment_branch_policy": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"protected_branches": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"custom_branch_policies": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceGithubRepositoryEnvironmentCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Owner).v3client
+
+	owner := meta.(*Owner).name
+	repoName := d.Get("repository").(string)
+	envName := d.Get("environment").(string)
+	updateData := createUpdateEnvironmentData(d, meta)
+
+	ctx := context.Background()
+
+	log.Printf("[DEBUG] Creating repository environment: %s/%s/%s", owner, repoName, envName)
+	_, _, err := client.Repositories.CreateUpdateEnvironment(ctx, owner, repoName, envName, &updateData)
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(buildTwoPartID(repoName, envName))
+
+	return resourceGithubRepositoryEnvironmentRead(d, meta)
+}
+
+func resourceGithubRepositoryEnvironmentRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*Owner).v3client
+
+	owner := meta.(*Owner).name
+	repoName, envName, err := parseTwoPartID(d.Id(), "repository", "environment")
+	if err != nil {
+		return err
+	}
+
+	ctx := context.WithValue(context.Background(), ctxId, d.Id())
+
+	log.Printf("[DEBUG] Reading repository environment: %s (%s/%s/%s)", d.Id(), owner, repoName, envName)
+	env, _, err := client.Repositories.GetEnvironment(ctx, owner, repoName, envName)
+	if err != nil {
+		if ghErr, ok := err.(*github.ErrorResponse); ok {
+			if ghErr.Response.StatusCode == http.StatusNotModified {
+				return nil
+			}
```

Comment:
nit: do we need this handling?  Surprised that a `3xx` is surfaced as an error.  
👌🏾 to keep but threw us off while reviewing.